### PR TITLE
Run WeatherService in a separate process implementation

### PR DESCRIPTION
**Pull request by MarsLviv**

### CODE REVIEW

File: app/src/main/java/com/example/weather/WeatherRequest.java [hunk 1]
```diff
@@ -28,6 +30,46 @@
...
+    public int describeContents() {
+        return 0;
+    }
+
+    @Override
+    public void writeToParcel(Parcel parcel, int flags) {
+        parcel.writeString(weatherApiKey);
+        parcel.writeString(weatherApiEntryPoint);
+        parcel.writeString(cityName);
+        parcel.writeParcelable(location, flags);
```
Comment (MarsLviv) [inline app/src/main/java/com/example/weather/WeatherRequest.java:53]: Please, help me deal with issue in this place.
Parcel object Location in this way cause null object reference in MainActivity with IWeatherServiceCallback.
If parcel object Location as commented in lines 55-58 - without any error.

Comment (MarsLviv) [reply app/src/main/java/com/example/weather/WeatherRequest.java:53]: thanks

File: app/src/main/java/com/example/weather/MainActivity.java [hunk 2]
```diff
@@ -84,26 +88,44 @@ private void onClickByCity(View view) {
...
+                        @Override
+                        public void run() {
+                            showWeatherView.setText(getText(R.string.default_weather_message).toString());
+                            Toast.makeText(MainActivity.this, getText(R.string.error_wrong_request).toString(), Toast.LENGTH_SHORT).show();
+                            weatherByCityButton.setEnabled(true);
+                        }
+                    });
+                    Log.i(TAG, "onClickByCity: onError");
+                }
+            });
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/MainActivity.java]: use lambdas everywhere where it is possible

File: app/src/main/java/com/example/weather/WeatherInfo.java [hunk 3]
```diff
@@ -15,6 +19,17 @@
         this.windSpeed = windSpeed;
     }
 
+    private WeatherInfo(Parcel parcel) {
+        readFromParcel(parcel);
+    }
+
+    public void readFromParcel(Parcel parcel) {
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/WeatherInfo.java]:
```suggestion
    private void readFromParcel(Parcel parcel) {
```

File: app/src/main/java/com/example/weather/WeatherService.java [hunk 4]
```diff
@@ -14,66 +14,67 @@
...
 
-    private final IBinder binder = new WeatherBinder();
-
-    private Handler handler;
+    IWeatherService.Stub binder = new IWeatherService.Stub() {
+        @Override
+        public void getCurrentWeatherInfo(WeatherRequest weatherRequest, IWeatherServiceCallback callback) {
+            Log.i(TAG, "make binder");
+            WeatherProviderRunnable weatherProviderRunnable = new WeatherProviderRunnable(weatherRequest, callback);
+            executorService.execute(weatherProviderRunnable);
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/WeatherService.java]: Move out this code to the Service level method.
```suggestion
            WeatherService.this.getCurrentWeatherInfo(request, callback);
```

File: app/src/main/java/com/example/weather/WeatherService.java [hunk 5]
```diff
@@ -14,66 +14,67 @@
...
             WeatherInfoProvider weatherInfoProvider = new WeatherInfoProvider(weatherRequest);
             weatherInfoProvider.provideWeather(new WeatherInfoProvider.RequestCallback() {
                 @Override
                 public void onRequestSucceed(@NonNull WeatherInfo weatherInfo) {
-                    handler.post(() -> callback.onWeatherInfoObtained(weatherInfo));
+                    try {
+                        callback.onWeatherInfoObtained(weatherInfo);
+                    } catch (RemoteException e) {
+                        e.printStackTrace();
+                        Log.i(TAG, "WeatherProviderRunnable: onRequestSucceed: couldn't make RequestCallback");
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/WeatherService.java]:
```suggestion
                        Log.e(TAG, "onRequestSucceed: got RemoteException while notifying the client", e);
```

File: app/src/main/java/com/example/weather/WeatherService.java [hunk 6]
```diff
@@ -14,66 +14,67 @@
...
                 }
 
                 @Override
                 public void onRequestFailed() {
-                    handler.post(() -> callback.onError());
+                    try {
+                        callback.onError();
+                    } catch (RemoteException e) {
+                        e.printStackTrace();
+                        Log.i(TAG, "WeatherProviderRunnable: onRequestFailed: couldn't make RequestCallback");
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/WeatherService.java]:
```suggestion
                        Log.e(TAG, "onRequestFailed: got RemoteException while notifying the client", e);
```

File: app/src/main/java/com/example/weather/MainActivity.java [hunk 7]
```diff
@@ -113,25 +135,43 @@ public void onClickByLocation(View view) {
...
+                                    weatherByLocationButton.setEnabled(true);
+                                }
+                            });
+                            Log.i(TAG, "onClickByLocation: onError");
+                        }
+                    });
+                } catch (RemoteException e) {
+                    e.printStackTrace();
+                    Log.i(TAG, "onClickByLocation: RemoteException during weather service request");
+                }
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/MainActivity.java]: Try to define single IWeatherServiceCallback.Stub as class field and use it while calling getCurrentWeather().

File: app/src/main/java/com/example/weather/MainActivity.java [hunk 8]
```diff
@@ -62,6 +50,40 @@ protected void onCreate(Bundle savedInstanceState) {
 
         locationClient = new LocationClient(this);
 
+        weatherServiceCallback = new IWeatherServiceCallback.Stub() {
```
Comment (gutachkee) [inline app/src/main/java/com/example/weather/MainActivity.java]: no need to have those inner classes declaration/field initialization in onCreate(). move out to the field initialization.

**Review by gutachkee — APPROVED** [review]
one more comment. Everything else looks good.

File: app/src/main/java/com/example/weather/MainActivity.java (hunk 8 above)
Comment (MarsLviv) [reply app/src/main/java/com/example/weather/MainActivity.java]: 1. weatherServiceCallback initialization I moved out to declaration section.
2. locationClient constructor has input parameter reference to MainActivity, and in time of locationClient's declaration MainActivity still not constructed. Therefore locationClient's  initializatin are in onCreate method.

**Pull request merged**